coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java

```diff
@@ -1027,42 +1013,78 @@ private void append(
+                                recordToAppend.headers()
                             );
+
+                            if (!hasRoomFor) {
+                                // If flushing fails, we don't catch the exception in order to let
+                                // the caller fail the current operation.
+                                flushCurrentBatch();
+                                maybeAllocateNewBatch(
+                                    producerId,
+                                    producerEpoch,
+                                    verificationGuard,
+                                    currentTimeMs
+                                );
+                            }
                         }
 
-                        currentBatch.builder.append(recordToAppend);
-                        currentBatch.nextOffset++;
-                    } catch (Throwable t) {
-                        log.error("Replaying record {} to {} failed due to: {}.", recordToReplay, tp, t.getMessage());
+                        try {
+                            if (replay) {
+                                coordinator.replay(
+                                    currentBatch.nextOffset,
+                                    producerId,
+                                    producerEpoch,
+                                    recordToReplay
+                                );
+                            }
+
+                            currentBatch.builder.append(recordToAppend);
+                            currentBatch.nextOffset++;
+                        } catch (Throwable t) {
+                            log.error("Replaying record {} to {} failed due to: {}.", recordToReplay, tp, t.getMessage());
 
-                        // Add the event to the list of pending events associated with the last
-                        // batch in order to fail it too.
-                        currentBatch.deferredEvents.add(event);
+                            // Add the event to the list of pending events associated with the last
+                            // batch in order to fail it too.
+                            currentBatch.deferredEvents.add(event);
 
-                        // If an exception is thrown, we fail the entire batch. Exceptions should be
-                        // really exceptional in this code path and they would usually be the results
-                        // of bugs preventing records to be replayed.
-                        failCurrentBatch(t);
+                            // If an exception is thrown, we fail the entire batch. Exceptions should be
+                            // really exceptional in this code path and they would usually be the results
+                            // of bugs preventing records to be replayed.
+                            failCurrentBatch(t);
 
-                        return;
+                            return;
+                        }
```

Comment:
It's kind of ugly to have a nested try-catch for the same exceptions. But I think the more specific error log for replay errors is useful?